dlang/source/flatbuffers/flatbufferbuilder.d

```diff
@@ -0,0 +1,308 @@
+	{
+		//You should not be creating any other objects or strings/vectors
+		//while an object is being constructed.
+		if(_vtable)
+			throw new Exception("FlatBuffers: object serialization must not be nested.");
+	}
+	
+	void startObject(int numfields)
+	{
+		notNested();
+		_vtable = new int[](numfields);
+		_objectStart = offset();
+	}
+	
+	///Set the current vtable at `voffset` to the current location in the buffer.
+	void slot(int voffset)
+	{
+		_vtable[voffset] = offset();
+	}
+	
+	///Add a scalar to a table at `o` into its vtable, with value `x` and default `d`.
+	void addBool(int o, bool x, bool d) { if(x != d) { addBool(x); slot(o); } }
+	void addByte(int o, byte x, byte d) { if(x != d) { addByte(x); slot(o); } }
+	void addUbyte(int o, ubyte x, ubyte d) { if(x != d) { addUbyte(x); slot(o); } }
+	void addShort(int o, short x, int d) { if(x != d) { addShort(x); slot(o); } }
+	void addUshort(int o, ushort x, ushort d) { if(x != d) { addUshort(x); slot(o); } }
+	void addInt(int o, int x, int d) { if(x != d) { addInt(x); slot(o); } }
+	void addUint(int o, uint x, uint d) { if(x != d) { addUint(x); slot(o); } }
+	void addLong(int o, long x, long d) { if(x != d) { addLong(x); slot(o); } }
+	void addUlong(int o, ulong x, ulong d) { if(x != d) { addUlong(x); slot(o); } }
+	void addFloat(int o, float x, double d) { if(x != d) { addFloat(x); slot(o); } }
+	void addDouble(int o, double x, double d) { if(x != d) { addDouble(x); slot(o); } }
+	void addOffset(int o, int x, int d) { if(x != d) { addOffset(x); slot(o); } }
+	
+	int createString(string s)
+	{
+		notNested();
+		auto utf8 = cast(ubyte[])s;
+		addUbyte(cast(ubyte)0);
+		startVector(1, cast(int)utf8.length, 1);
+		_space -= utf8.length;
+		_buffer.data[_space.._space + utf8.length] = utf8[];
+		return endVector();
+	}
+	
+	///Structs are stored inline, so nothing additional is being added.
+	///`d` is always 0.
+	void addStruct(int voffset, int x, int d)
+	{
+		if(x != d)
+		{
+			nested(x);
+			slot(voffset);
+		}
+	}
+	
+	int endObject()
+	{
+		if(!_vtable)
+			throw new InvalidOperationException("Flatbuffers: calling endObject without a startObject");
+		
+		addInt(cast(int)0);
+		auto vtableloc = offset();
+		
+		//Write out the current vtable.
+		for(int i=cast(int)_vtable.length-1; i>=0; i--)
+		{
+			//Offset relative to the start of the table.
+			short off = cast(short)(_vtable[i] != 0? vtableloc - _vtable[i] : 0);
+			addShort(off);
+		}
+		
+		const int standardFields = 2; //The fields below:
+		addShort(cast(short)(vtableloc - _objectStart));
+		addShort(cast(short)((_vtable.length + standardFields) * short.sizeof));
+		
+		///Search for an existing vtable that matches the current one.
+		int existingVtable = 0;
+		
+	frist: for(int i=0; i<_numVtables; i++)
+		{
+			int vt1 = _buffer.length - _vtables[i];
+			int vt2 = _space;
+			short len = _buffer.get!short(vt1);
+			if(len == _buffer.get!short(vt2))
+			{
+				for(int j=short.sizeof; j<len; j+=short.sizeof)
```

Comment:
is there no D equivalent of memcmp you can add to bytebuffer?

Reply:
FYI: `x[] = y[];` can be used to copy content of y into x.

This works for dynamic arrays and static arrays.

https://dlang.org/spec/arrays.html#array-copying

For untyped pieces of memory, you can cast it into `short[]`  for example quite easily:  `short[] x = buffer[0 .. len];`. Then use slice / array `x` as you would normally use it in D.


For unstructured data, you can use D implementation of `memcpy `:

https://dlang.org/library/core/stdc/string/memcpy.html

or just declare `extern(C) memcpy...` from C and call it the same.